docs/proposal-architecture.md

```diff
@@ -0,0 +1,111 @@
+- Detailed explaination for using multiple cores is in "Worker Proposal" document.
+#### Step5: Relocating
+- Master Node pick two Worker Nodes and change its block until there is nothing to relocate.
+    - block: the unit of moving data.
+- Detailed explaination is in "Master Proposal" document.
+- Example for brief understaning of Step5.
+    - Suppose there are three Worker Nodes named A, B, and C.
+        - Notation
+            - Worker name's meta variable is X, Y.
+            - R[X]: Range of each Worker Node X.
+            - D[X,Y]: Rough sorted data of range R[Y] in X. 
+    - 1. Master Node pick A and B.
+    - 2. Then A gives one block in D[A,B] to Master Node.
+    - 3. Then B gives one block in D[B,A] to Master Node.
+    - 4. Do this until there is no block to relocate for both A and B.
+    - 5. Master Node pick A and C, then do 2~4. 
+    - 6. Master Node pick B and C, then do 2~4. 
+
+### Second Phase
+
+After data relocation, each Worker Node has its local data organized into partitions. In the second phase, all we have to do is just sorting the data within each partition. This automatically lead to overall sorting. 
+
+#### Step1: Local Parallel Sorting
+- Each Worker does merge sort in parallel with respect to other Worker.
+- Detailed explaination for using multiple cores is in "Worker Proposal" document.
+
+#### Step2: Singal Sorting Completion to Master
+- Upon completing the sort for all partitions, each Worker Node sends a signal to the Master Node indicating that it has finished sorting.
+- The Master Node waits until all Worker Nodes gives signal to it.
+- If all Worker Node gives signal to Master Node, program ends.
+
+## More Detail
+Until now, we got a big picture of achieving Goal, mainly about answering the question of "How to corrodinate Multiple Nodes?". The four Proposal document below will give you detailed explaination for each subject.
+
+### Data Structure Proposal
+- Define fundamental data structure that we will use through this project.
+### RPC Protocol Proposal
+- Data Sample Request and Reply
+    - Master sends a request to each Worker.
+    - Each Worker replies with a set of sample keys.
+- Key Range Broadcast
+    - Master broadcasts the key ranges to all Worker Nodes.
+- Relocation Request and Reply
+    - Master sends a request for data block movement.
+    - Worker sends a reply the data block that will relocated.
+- Sorting Complete Signal
+    - Worker sends a signal to the Master.
+    - Master keeps a count and waits for signals from all Workers.
+### Worker Proposal
+- Data Sampling
+    - On request from the Master Node, the Worker sends back a small subset of records.
+- Key Range Partitioning
+    - Upon receiving key ranges, partition data based on those ranges.
+    - Use parallelism
+- Local Sorting 
+    - Perform merge sort on each partition.
+    - Use parallelism
+- Signalling
+    - Once sorting is done, send a signal to the Master Node.
+### Master Proposal
```

Comment:
Master node could become a potential bottleneck and a single point of failure in the system. Therefore, I think it's essential to discuss strategies to mitigate this problem.

Reply:
Agreed. However, if all algorithms are deterministic based on state (e.g. blocks), it is okay to re-run all phases again. Therefore, naively, we can re-run all of things if failure occured.